extension/src/time_weighted_average.rs

```diff
@@ -0,0 +1,422 @@
+
+use serde::{Serialize, Deserialize};
+use pg_sys::TimestampTz;
+use pg_sys::Datum;
+use std::slice;
+
+use flat_serialize::*;
+use pgx::*;
+use crate::{
+    aggregate_utils::{aggregate_mctx, in_aggregate_context},
+    flatten,
+    json_inout_funcs,
+    palloc::{Internal, in_memory_context}, pg_type
+};
+
+use time_weighted_average::{
+    tspoint::TSPoint, 
+    TimeWeightSummary,
+    TimeWeightError,
+    TimeWeightMethod,
+};
+
+// hack to allow us to qualify names with "timescale_analytics_experimental"
+// so that pgx generates the correct SQL
+mod timescale_analytics_experimental {
+    pub(crate) use super::*;
+    extension_sql!(r#"
+        CREATE SCHEMA IF NOT EXISTS timescale_analytics_experimental;
+    "#);
+}
+extension_sql!(r#"set search_path to 'timescale_analytics_experimental', 'public';"#);
+
+// This assumes ordered input and will not sort interally, will error if input is unsorted.
+#[pg_extern(schema = "timescale_analytics_experimental")]
+pub fn time_weight_ordered_trans(
+    state: Option<Internal<TimeWeightSummary>>,
+    method: String,
+    ts: Option<pg_sys::TimestampTz>,
+    val: Option<f64>,
+    fcinfo: pg_sys::FunctionCallInfo,
+) -> Option<Internal<TimeWeightSummary>> {
+    unsafe {
+        in_aggregate_context(fcinfo, || {
+            let p = match (ts, val) {
+                (_, None) => return state,
+                (None, _) => return state, 
+                (Some(ts), Some(val)) => TSPoint{ts, val},
+            };
+            let method = match method.to_lowercase().as_str() {
+                "linear"=> TimeWeightMethod::Linear,
+                "locf" => TimeWeightMethod::LOCF,
+                &_ => panic!()
+            };
```

Comment:
I think this should be moved into the `None` branch of the `match state`; no reason to reparse after the first time

Reply:
Hmm...is there a way to mark a function input that way in an aggregate? I could imagine weird cases where you'd have this change, but they're all very odd and would lead to weird errors, ie a case statement in the select that coudl give odd results if you're not careful...